### PR TITLE
feat: added pre/post conditions for lender system tests, unhappy cases, new lender config setup

### PR DESCRIPTION
**Pull request by charleenfei**
* creates BaseDistributorFab
* whitelistinvestor -> investor, move to users
* updates tests for pre/post conditions
* optimises lender deploy config/setup

### CODE REVIEW

File: src/core/lender/deployer.sol [hunk 1]
```diff
@@ -23,7 +23,9 @@ import {WhitelistOperator} from "./tranche/operator/whitelist.sol";
 import {Tranche} from "./tranche/tranche.sol";
 import {SeniorTranche} from "./tranche/senior_tranche.sol";
 import {SwitchableDistributor} from "./distributor/switchable.sol";
+import {BaseDistributor} from "./distributor/base.sol";
 import "tinlake-erc20/erc20.sol";
+import "../test/system/interfaces.sol";
```
Comment (manuelpolzhofer) [inline src/core/lender/deployer.sol]: never include test-code in the actual code base

File: src/core/lender/deployer.sol [hunk 2]
```diff
@@ -90,23 +91,16 @@ contract SwitchableDistributorFab {
     }
 }
 
-// todo add BaseDistributorFab
-
-contract DistributorLike {
```
Comment (manuelpolzhofer) [inline src/core/lender/deployer.sol]: we should keep the `DistributorLike` and `OperatorLike` here

File: src/core/lender/deployer.sol [hunk 3]
```diff
@@ -80,7 +81,7 @@ contract DistributorFab {
     function newDistributor(address currency) public returns (address);
 }
 
-contract SwitchableDistributorFab {
+contract SwitchableFab {
```
Comment (manuelpolzhofer) [inline src/core/lender/deployer.sol]: I think we should keep the pattern.

`ContractName+Fab` therefore it should be `SwitchableDistributorFab`

File: src/core/lender/deployer.sol [hunk 4]
```diff
@@ -90,23 +91,16 @@ contract SwitchableDistributorFab {
...
-// todo add BaseDistributorFab
-
-contract DistributorLike {
-    function rely(address usr) public;
-    function deny(address usr) public;
-    function depend (bytes32 what, address addr) public;
-    function file(bytes32 what, bool flag) public;
-    function balance() public;
-}
+contract BaseFab {
```
Comment (manuelpolzhofer) [inline src/core/lender/deployer.sol]: same here

File: src/core/test/system/borrower/issue.t.sol [hunk 5]
```diff
@@ -24,7 +24,7 @@ contract IssueTest is SystemTest {
     address borrower_;
         
     function setUp() public {
-        baseSetup();
+        baseSetup("whitelist", "switchable");
```
Comment (manuelpolzhofer) [inline src/core/test/system/borrower/issue.t.sol]: I would prefer a variable here:
```javascript
bytes32 juniorOperator = "whitelist";
bytes32 distributor = "switchable";
baseSetup(juniorOperator, distributor);
```

File: src/core/test/system/borrower/lock.sol [hunk 6]
```diff
@@ -24,7 +24,7 @@ contract LockTest is SystemTest {
     address borrower_;
         
     function setUp() public {
-        baseSetup();
+        baseSetup("whitelist", "switchable");
```
Comment (manuelpolzhofer) [inline src/core/test/system/borrower/lock.sol]: same here

File: src/core/test/system/interfaces.sol [hunk 7]
```diff
@@ -58,4 +58,33 @@ contract ERC20Like {
 contract CeilingLike {
     function values(uint) public view returns(uint);
     function file(uint loan, uint principal) public;
+}
+
+contract OperatorLike {
```
Comment (manuelpolzhofer) [inline src/core/test/system/interfaces.sol]: rename it here

File: src/core/test/system/interfaces.sol [hunk 8]
```diff
@@ -58,4 +58,33 @@ contract ERC20Like {
...
+contract TrancheLike {
+    function balance() public returns(uint);
+    function tokenSupply() public returns(uint);
+}
+
+contract AssessorLike {
+    function calcTokenPrice(address tranche) public returns(uint);
+}
+
+contract DistributorLike {
```
Comment (manuelpolzhofer) [inline src/core/test/system/interfaces.sol]: rename it here

File: src/core/test/system/lender/redeem.t.sol [hunk 9]
```diff
@@ -20,26 +20,67 @@ import {SwitchableDistributor} from "../../../lender/distributor/switchable.sol"
...
     }
 
     function supply(uint balance, uint amount) public {
         currency.mint(juniorInvestor_, balance);
         juniorInvestor.doSupply(amount);
     }
-
-    function testRedeem() public {
+    
+    function testSwitchableRedeem() public {
```
Comment (manuelpolzhofer) [inline src/core/test/system/lender/redeem.t.sol]: 👍

File: src/core/lender/deployer.sol (hunk 3 above)
Comment (charleenfei) [reply src/core/lender/deployer.sol]: We have this for the Allowance and Whitelist Operators though (the naming pattern)

File: src/core/lender/deployer.sol (hunk 4 above)
Comment (ilinzweilin) [reply src/core/lender/deployer.sol]: @charleenfei I agree with Manuel BaseFab sounds really generic

File: src/core/test/system/lender/redeem.t.sol [hunk 10]
```diff
@@ -20,26 +20,69 @@ import {SwitchableDistributor} from "../../../lender/distributor/switchable.sol"
 
 contract RedeemTest is SystemTest {
```
Comment (ilinzweilin) [inline src/core/test/system/lender/redeem.t.sol]: maybe create a switchable_distributor folder in lende rand put all the related tests there

**Pull request merged**